### PR TITLE
Issue 4896: Adding new Integration tests for Feature Key Value Store

### LINKED ISSUE #4896: Add new Integration tests for Key Value Store feature
**Problem description**
Add System test for Key Value Store feature

**Problem location**
io.pravega.test.system

**Suggestions for an improvement**

### PR DESCRIPTION
**Pull request by SomeshJoshi19**
**Change log description**  
This PR has system test to verify single & multiple table entries over conditional update and removal operations along with delete KVT test.

**Purpose of the change**  
Fixes #4896 

**What the code does**  
* Adds tests to verify insert same key under mutiple KF: kvt.put()
* conditional insert if key does not exists:  kvt.putIfAbsent()
* conditional update KVP: kvt.replace()
* conditional delete KVP: kvt.remove()
* list key version set: getVersion()
* conditional insert multiple Table Entries: kvt.replaceAll()
* conditional update multiple Table Entries: kvt.replaceAll()
* conditiontional delete multiple Table Entries: kvt.removeAll()
* retrieve KVP after deleting KVT: manger.deleteKeyValueTable()

**How to verify it**  
All the tests present under this file must pass

### CODE REVIEW

File: test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java [hunk 1]
```diff
@@ -0,0 +1,744 @@
...
+    private Service controllerInstance;
+
+    @Before
+    public void setup() throws Exception {
+        controllerInstance = Utils.createPravegaControllerService(null);
+        List<URI> ctlURIs = controllerInstance.getServiceDetails();
+        final List<String> uris = ctlURIs.stream().filter(ISGRPC).map(URI::getAuthority).collect(Collectors.toList());
+        controllerURI = URI.create("tcp://" + String.join(",", uris));
+        final ClientConfig clientConfig = Utils.buildClientConfig(controllerURI);
+        final ScheduledExecutorService controllerExecutor= Executors.newScheduledThreadPool(5);
```
Comment (andreipaduroiu) [inline test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java:58]: 1. use ExecutorServiceHelpers.newScheduledThreadPool.
2. store this in a class variable and shut it down in `tearDown` (`ExecutorServiceHelpers.shutdown`)

File: test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java [hunk 2]
```diff
@@ -0,0 +1,744 @@
...
+
+    @After
+    public void tearDown() {
+        this.connectionFactory.close();
+        this.controller.close();
+        this.keyValueTableFactory.close();
+        this.keyValueTable.close();
+    }
+
+    // Test case-42: Insert Same Key into multiple KeyFamilies
```
Comment (andreipaduroiu) [inline test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java:81]: use Javadoc

File: test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java [hunk 3]
```diff
@@ -0,0 +1,744 @@
...
+    @Test
+    public void test1SameKeyMutipleKF() throws UnsupportedOperationException {
```
Comment (andreipaduroiu) [inline test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java:83]: no need for this `throws` declaration.

File: test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java [hunk 4]
```diff
@@ -0,0 +1,744 @@
...
+            assertFalse("Corresponding values should be different for both the entries",
+                    (tableEntry1.getValue()).equals(tableEntry2.getValue()));
+            log.info("Corresponding values are different for both the entries '{}', '{}'",
+                    tableEntry1.getValue(), tableEntry2.getValue());
+
+            log.info("Table entry1, key: {} and value: {}", tableEntry1.getKey(), tableEntry1.getValue());
+            log.info("Table entry2, key: {} and value: {}", tableEntry2.getKey(), tableEntry2.getValue());
+            log.info("Successfully completed test Insert Same Key into multiple KeyFamilies");
+        }
+        catch (UnsupportedOperationException ex) {
```
Comment (andreipaduroiu) [inline test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java:117]: This is unnecessary

File: test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java [hunk 5]
```diff
@@ -0,0 +1,744 @@
...
+            throw new UnsupportedOperationException();
+        }
+        catch (Exception e) {
```
Comment (andreipaduroiu) [inline test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java:120]: Remove this.

File: test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java (hunk 2 above)
Comment (andreipaduroiu) [reply test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java:81]: Fix this here and for all methods below. Same request applies for all other comments I made.

File: test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java [hunk 6]
```diff
@@ -0,0 +1,744 @@
...
+            log.info("Version2: {}", version2);
+
+            TableEntry<Integer, String> tableEntry1 = this.keyValueTable.get(kf1, key).join();
+            TableEntry<Integer, String> tableEntry2 = this.keyValueTable.get(kf2, key).join();
+
+            assertEquals("Keys should be same for both entries",
+                    tableEntry1.getKey().getKey(), tableEntry2.getKey().getKey());
+            assertTrue("Corresponding value for entry1 should be as inserted",
+                    value1.equals(tableEntry1.getValue()));
+            log.info("Corresponding value '{}' for entry1 is inserted '{}'", value1, tableEntry1.getValue());
```
Comment (andreipaduroiu) [inline test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java:104]: What's the point of this log.info? You have already validated that the two values are equal so why log it?

File: test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java [hunk 7]
```diff
@@ -0,0 +1,744 @@
...
+            Version version2 = this.keyValueTable.put(kf2, key, value2).join();
+            log.info("Version1: {}", version1);
+            log.info("Version2: {}", version2);
+
+            TableEntry<Integer, String> tableEntry1 = this.keyValueTable.get(kf1, key).join();
+            TableEntry<Integer, String> tableEntry2 = this.keyValueTable.get(kf2, key).join();
+
+            assertEquals("Keys should be same for both entries",
+                    tableEntry1.getKey().getKey(), tableEntry2.getKey().getKey());
+            assertTrue("Corresponding value for entry1 should be as inserted",
```
Comment (andreipaduroiu) [inline test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java:102]: assertEquals

File: test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java [hunk 8]
```diff
@@ -0,0 +1,744 @@
...
+                    value1.equals(tableEntry1.getValue()));
+            log.info("Corresponding value '{}' for entry1 is inserted '{}'", value1, tableEntry1.getValue());
+            assertTrue("Corresponding value for entry1 should be as inserted",
```
Comment (andreipaduroiu) [inline test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java:105]: assertEquals

File: test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java [hunk 9]
```diff
@@ -0,0 +1,744 @@
...
+                    value2.equals(tableEntry2.getValue()));
+            log.info("Corresponding value '{}' for entry2 is inserted '{}'", value2, tableEntry2.getValue());
```
Comment (andreipaduroiu) [inline test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java:107]: superfluous.

File: test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java [hunk 10]
```diff
@@ -0,0 +1,744 @@
...
+            assertFalse("Corresponding values should be different for both the entries",
+                    (tableEntry1.getValue()).equals(tableEntry2.getValue()));
+            log.info("Corresponding values are different for both the entries '{}', '{}'",
```
Comment (andreipaduroiu) [inline test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java:110]: here too

File: test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java [hunk 11]
```diff
@@ -0,0 +1,744 @@
...
+        }
+        catch (UnsupportedOperationException ex) {
+            throw new UnsupportedOperationException();
+        }
+        catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    // Test case-43: Conditional insert KVP, if key not exists already
```
Comment (andreipaduroiu) [inline test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java:125]: Same comments in this method as above.

File: test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java [hunk 12]
```diff
@@ -0,0 +1,744 @@
...
+            throw new UnsupportedOperationException();
+        }
+        catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    // Test case-44: Conditional update Key value using its version
+    @Test
+    public void test3ConditionUpdateKeyValue() throws UnsupportedOperationException {
```
Comment (andreipaduroiu) [inline test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java:168]: Same comments here too.

File: test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java [hunk 13]
```diff
@@ -0,0 +1,744 @@
...
+            throw new UnsupportedOperationException();
+        }
+        catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    // Test case-45: Conditional delete single Key using its version
+    @Test
+    public void test4ConditionDeleteKey() throws UnsupportedOperationException {
```
Comment (andreipaduroiu) [inline test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java:254]: And here, and the remaining ones below.

File: test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java [hunk 14]
```diff
@@ -0,0 +1,744 @@
...
+        }
+        catch (UnsupportedOperationException ex) {
+            throw new UnsupportedOperationException();
+        }
+        catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    public Integer getKeyID() {
```
Comment (andreipaduroiu) [inline test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java:718]: private

File: test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java [hunk 15]
```diff
@@ -0,0 +1,744 @@
...
+        Integer keyId = ThreadLocalRandom.current().nextInt(1000,99999);
```
Comment (andreipaduroiu) [inline test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java:719]: create a class-level variable of type `Random`, seed it with a **constant**, and use that for generating random numbers. For reproducibility, it is important to use the same inputs.

File: test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java [hunk 16]
```diff
@@ -0,0 +1,744 @@
...
+        return keyId;
+    }
+
+    public String getValue() {
```
Comment (andreipaduroiu) [inline test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java:723]: private

File: test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java [hunk 17]
```diff
@@ -0,0 +1,744 @@
...
+        String value = "Hello World";
```
Comment (andreipaduroiu) [inline test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java:724]: Consider a more efficient way of generating a value. You are executing the loop around 2500 times, in each case allocating more and more memory.

File: test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java [hunk 18]
```diff
@@ -0,0 +1,744 @@
...
+        log.debug("Initial value length: {}", value.getBytes().length);
+        while (value.getBytes().length <= 10302) {
+           value+= ThreadLocalRandom.current().nextInt(1000, 9999);
+        }
+        log.debug("Final value length: {}", value.getBytes().length);
```
Comment (andreipaduroiu) [inline test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java:729]: Remove all this logging.

File: test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java [hunk 19]
```diff
@@ -0,0 +1,744 @@
...
+        return value;
+    }
+
+    private static class IntegerSerializer implements Serializer<Integer> {
```
Comment (andreipaduroiu) [inline test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java:733]: This implementation is incorrect and is not needed. Use `io.pravega.test.integration.utils.IntegerSerializer`.

File: test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java (hunk 1 above)
Comment (SomeshJoshi19) [reply test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java:58]: Sure, will change as said.

File: test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java (hunk 2 above)
Comment (SomeshJoshi19) [reply test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java:81]: Sure, will change as suggested.

File: test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java (hunk 6 above)
Comment (SomeshJoshi19) [reply test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java:104]: Just added to verify the logs if it passes the assert check. Will remove now

File: test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java (hunk 15 above)
Comment (SomeshJoshi19) [reply test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java:719]: okay, will do modify it the same way.

File: test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java (hunk 19 above)
Comment (SomeshJoshi19) [reply test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java:733]: okay, will be using this instead as suggested: 
```
private static class IntegerSerializer implements Serializer<Integer> {
        @Override
        public ByteBuffer serialize(Integer value) {
            ByteBuffer result = ByteBuffer.allocate(4).putInt(value);
            result.rewind();
            return result;
        }

        @Override
        public Integer deserialize(ByteBuffer serializedValue) {
            return serializedValue.getInt();
        }
    }
```
Is it correct ?

File: test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java (hunk 3 above)
Comment (SomeshJoshi19) [reply test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java:83]: okay, I have removed it now.

File: test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java (hunk 4 above)
Comment (SomeshJoshi19) [reply test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java:117]: okay, I have removed it now.

File: test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java (hunk 18 above)
Comment (SomeshJoshi19) [reply test/system/src/test/java/io/pravega/test/system/KeyValueStoreTest.java:729]: Sure, removed.

**Comment by SomeshJoshi19** [comment]
Closing this PR as there are many merge conflicts in the source branch.

**Pull request closed without merging**